### PR TITLE
Changes the build system to CMake

### PR DESCRIPTION
**Pull request by ilumsden**
Since @hariharan-devarajan mentioned that we wanted to do this and since I had some free time on the flight to SC, I just went ahead and started implementing a CMake-based build system for DYAD.

As of now, this PR just implements a drop-in replacement for autotools. It does **not** restructure the code base, add any major changes to the code, etc.

In case anyone else wants to contribute to this, I put the branch directly on this repo (as opposed to a fork).

### CODE REVIEW

**Comment by hariharan-devarajan** [comment]
@ilumsden I will be looking into this actively. I will look at your branch and continue this work from here.

**Comment by hariharan-devarajan** [comment]
Did the following changes.

- [ ] Enabled CMake based Build
- [ ] Disabled Python test as its hanging.

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+set(DYAD_DATA_PLANE "FLUX_RPC" CACHE STRING "Protocol to use for DYAD's Data Plane")
+set_property(CACHE DYAD_DATA_PLANE PROPERTY STRINGS FLUX_RPC UCX)
+set(DYAD_PROFILER "NONE" CACHE STRING "Protocol to use for DYAD's Data Plane")
+set_property(CACHE DYAD_PROFILER PROPERTY STRINGS PERFFLOW_ASPECT CALIPER NONE)
+
+#------------------------------------------------------------------------------
+# Build Type based configurations
+#------------------------------------------------------------------------------
+set(CMAKE_C_FLAGS ${CMAKE_C_FLAGS} -std=c99)
+set(CMAKE_CXX_FLAGS ${CMAKE_CXX_FLAGS} -std=c++11)
```
Comment (hariharan-devarajan) [inline CMakeLists.txt:93]: Use minimum version 11 not absolute

File: CMake/dyad-config.cmake.build.in [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    set(DYAD_FOUND FALSE)
+endif ()
+#message(STATUS "DYAD_INCLUDE_DIRS: " ${DYAD_INCLUDE_DIRS})
+get_filename_component(DYAD_ROOT_DIR ${DYAD_INCLUDE_DIRS}/.. ABSOLUTE)
+#message(STATUS "DYAD_ROOT_DIR: " ${DYAD_ROOT_DIR})
+set(DYAD_LIBRARY_PATH "@CMAKE_LIBRARY_OUTPUT_DIRECTORY@")
+#message(STATUS "DYAD_LIBRARY_PATH: " ${DYAD_LIBRARY_PATH})
+link_directories(${DYAD_LIBRARY_PATH})
+set(DYAD_LIBRARIES dlio-profiler)
+set(DYAD_DEFINITIONS "")
```
Comment (hariharan-devarajan) [inline CMake/dyad-config.cmake.build.in:15]: add alias dyad::dyad

File: CMake/dyad-config.cmake.build.in [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+include(FindPackageHandleStandardArgs)
+find_package_handle_standard_args(dlio-profiler
```
Comment (hariharan-devarajan) [inline CMake/dyad-config.cmake.build.in:17]: change to dyad

File: CMake/dyad-config.cmake.install.in [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+set(DYAD_INCLUDE_DIRS "@DYAD_INSTALL_INCLUDE_DIR@")
+if (NOT IS_DIRECTORY "${DYAD_INCLUDE_DIRS}")
+    set(DYAD_FOUND FALSE)
+endif ()
+#message(STATUS "DYAD_INCLUDE_DIRS: " ${DYAD_INCLUDE_DIRS})
+get_filename_component(DYAD_ROOT_DIR ${DYAD_INCLUDE_DIRS}/.. ABSOLUTE)
+#message(STATUS "DYAD_ROOT_DIR: " ${DYAD_ROOT_DIR})
+set(DYAD_LIBRARY_PATH "@DYAD_INSTALL_LIB_DIR@")
+link_directories(${DYAD_LIBRARY_PATH})
+set(DYAD_LIBRARIES dlio-profiler)
```
Comment (hariharan-devarajan) [inline CMake/dyad-config.cmake.install.in:13]: change to dyad

File: src/dyad/dtl/dyad_dtl_impl.c [hunk 5]
```diff
@@ -64,13 +65,14 @@ dyad_rc_t dyad_dtl_finalize (dyad_dtl_t **dtl_handle)
     } else if ((*dtl_handle)->mode == DYAD_DTL_FLUX_RPC) {
 #else
     if ((*dtl_handle)->mode == DYAD_DTL_FLUX_RPC) {
-#endif
+
```
Comment (hariharan-devarajan) [inline src/dyad/dtl/dyad_dtl_impl.c:68]: Check here

File: src/dyad/wrapper/flux_barrier.c [hunk 6]
```diff
@@ -92,12 +93,12 @@ int main (int argc, char** argv)
 
     flux_future_destroy (fb);
 
-    struct timespec t_now;
-    clock_gettime (CLOCK_REALTIME, &t_now);
-    char tbuf[100];
-    strftime (tbuf, sizeof (tbuf), "%D %T", gmtime (&(t_now.tv_sec)));
-
-    FLUX_LOG_INFO ("Synchronized at %s.%09ld\n", tbuf, t_now.tv_nsec);
+//    struct timespec t_now;
```
Comment (hariharan-devarajan) [inline src/dyad/wrapper/flux_barrier.c:96]: Check here. This was not compiling on GCC 11

File: CMakeLists.txt (hunk 1 above)
Comment (hariharan-devarajan) [reply CMakeLists.txt:93]: FIXED

File: CMake/dyad-config.cmake.build.in (hunk 2 above)
Comment (hariharan-devarajan) [reply CMake/dyad-config.cmake.build.in:15]: FIXED

File: CMake/dyad-config.cmake.build.in [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+if (NOT IS_DIRECTORY "${DYAD_INCLUDE_DIRS}")
+    set(DYAD_FOUND FALSE)
+endif ()
+#message(STATUS "DYAD_INCLUDE_DIRS: " ${DYAD_INCLUDE_DIRS})
+get_filename_component(DYAD_ROOT_DIR ${DYAD_INCLUDE_DIRS}/.. ABSOLUTE)
+#message(STATUS "DYAD_ROOT_DIR: " ${DYAD_ROOT_DIR})
+set(DYAD_LIBRARY_PATH "@CMAKE_LIBRARY_OUTPUT_DIRECTORY@")
+#message(STATUS "DYAD_LIBRARY_PATH: " ${DYAD_LIBRARY_PATH})
+link_directories(${DYAD_LIBRARY_PATH})
+set(DYAD_LIBRARIES dlio-profiler)
```
[the remark this answers is not in the document]
Comment (hariharan-devarajan) [reply CMake/dyad-config.cmake.build.in]: FIXED

File: CMake/dyad-config.cmake.install.in (hunk 4 above)
Comment (hariharan-devarajan) [reply CMake/dyad-config.cmake.install.in]: FIXED

**Comment by hariharan-devarajan** [comment]
@JaeseungYeom I fixed all the changes we talked about. Let me know if u think i should change anything else.

File: src/dyad/core/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,24 @@
+set(DYAD_CORE_SRC ${CMAKE_CURRENT_SOURCE_DIR}/dyad_core.c)
+set(DYAD_CORE_PRIVATE_HEADER ${CMAKE_CURRENT_SOURCE_DIR}/dyad_envs.h)
+set(DYAD_CORE_PUBLIC_HEADER ${CMAKE_CURRENT_SOURCE_DIR}/dyad_core.h)
```
Comment (ilumsden) [inline src/dyad/core/CMakeLists.txt:3]: If we don't want users directly calling DYAD Core, we should not have `dyad_core.h` in public headers. Alternatively, if we do want to install `dyad_core.h`, we need to add `dyad_dtl.h` to the public headers in the `dtl` directory.

File: src/dyad/dtl/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,43 @@
+# DTL Interface
+set(DTL_SRC ${CMAKE_CURRENT_SOURCE_DIR}/dyad_dtl_impl.c)
+set(DTL_PRIVATE_HEADERS ${CMAKE_CURRENT_SOURCE_DIR}/dyad_dtl_impl.h
+                        ${CMAKE_CURRENT_SOURCE_DIR}/dyad_rc.h)
+set(DTL_PUBLIC_HEADERS )
```
Comment (ilumsden) [inline src/dyad/dtl/CMakeLists.txt:5]: See the comment for `CMakeLists.txt` in `core`

File: src/dyad/core/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+add_library(dyad_core SHARED ${DYAD_CORE_SRC})
+target_link_libraries(dyad_core PUBLIC Jansson::Jansson flux::core)
+target_link_libraries(dyad_core PRIVATE utils murmur3 dyad_dtl)
```
Comment (ilumsden) [inline src/dyad/core/CMakeLists.txt:7]: In the `dtl` directory, the CMake target is named `${PROJECT_NAME}_dtl`. We should use the same here

File: src/dyad/modules/CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,42 @@
+set(DYAD_MODULE_SRC ${CMAKE_CURRENT_SOURCE_DIR}/dyad.c)
+set(DYAD_MODULE_PRIVATE_HEADERS )
+set(DYAD_MODULE_PUBLIC_HEADERS )
+
+add_library(dyad SHARED ${DYAD_MODULE_SRC})
+set_target_properties(dyad PROPERTIES PREFIX "")
+target_link_libraries(dyad PUBLIC Jansson::Jansson flux::core)
+target_link_libraries(dyad PRIVATE dyad_dtl)
```
Comment (ilumsden) [inline src/dyad/modules/CMakeLists.txt:8]: In the `dtl` directory, the CMake target is named `${PROJECT_NAME}_dtl`. We should use the same here

File: src/dyad/core/CMakeLists.txt (hunk 8 above)
Comment (hariharan-devarajan) [reply src/dyad/core/CMakeLists.txt:3]: Fixed

File: src/dyad/dtl/CMakeLists.txt (hunk 9 above)
Comment (hariharan-devarajan) [reply src/dyad/dtl/CMakeLists.txt:5]: We dont need it as core is not in public header

File: src/dyad/core/CMakeLists.txt (hunk 10 above)
Comment (hariharan-devarajan) [reply src/dyad/core/CMakeLists.txt]: Fixed

File: src/dyad/modules/CMakeLists.txt (hunk 11 above)
Comment (hariharan-devarajan) [reply src/dyad/modules/CMakeLists.txt]: Fixed

**Comment by JaeseungYeom** [comment]
Usually when I forgot to specify `CMAKE_INSTALL_PREFIX` with cmake command, I still make it work by editing the variable in `cmake_install.cmake` that is generated under build directory. However, I see `INSTALL DESTINATION` is fixed rather than depending on `CMAKE_INSTALL_PREFIX`. Can this behavior changed?

**Comment by JaeseungYeom** [comment]
Also, can you confirm the python test is working?

**Comment by hariharan-devarajan** [comment]
> Usually when I forgot to specify `CMAKE_INSTALL_PREFIX` with cmake command, I still make it work by editing the variable in `cmake_install.cmake` that is generated under build directory. However, I see `INSTALL DESTINATION` is fixed rather than depending on `CMAKE_INSTALL_PREFIX`. Can this behavior changed?

I fixed it can u check it?

**Comment by hariharan-devarajan** [comment]
> Also, can you confirm the python test is working?

It is working locally but not on CI.

**Comment by JaeseungYeom** [comment]
> > Usually when I forgot to specify `CMAKE_INSTALL_PREFIX` with cmake command, I still make it work by editing the variable in `cmake_install.cmake` that is generated under build directory. However, I see `INSTALL DESTINATION` is fixed rather than depending on `CMAKE_INSTALL_PREFIX`. Can this behavior changed?
> 
> I fixed it can u check it?

Yes. It works.

**Comment by JaeseungYeom** [comment]
Should python binding be installed as well?

File: CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+# This is verbose, maybe an alternate might help simplify
+#option(BUILD_URPC "Build DYAD's URPC code" OFF)
+#option(ENABLE_PERFFLOW "Build with PerfFlow Aspect support" OFF)
+#option(ENABLE_UCX_DTL "Build DYAD's UCX data transport layer" OFF)
+
+set(DYAD_CONTROL_PLANE "FLUX_RPC" CACHE STRING "Protocol to use for DYAD's Control Plane")
+set_property(CACHE DYAD_CONTROL_PLANE PROPERTY STRINGS FLUX_RPC)
+set(DYAD_DATA_PLANE "FLUX_RPC" CACHE STRING "Protocol to use for DYAD's Data Plane")
+set_property(CACHE DYAD_DATA_PLANE PROPERTY STRINGS FLUX_RPC UCX)
+set(DYAD_PROFILER "NONE" CACHE STRING "Protocol to use for DYAD's Data Plane")
```
Comment (JaeseungYeom) [inline CMakeLists.txt:69]: description of the option is irrelevant.

File: CMakeLists.txt [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+set_property(CACHE DYAD_PROFILER PROPERTY STRINGS PERFFLOW_ASPECT CALIPER NONE)
+
+#------------------------------------------------------------------------------
+# Build Type based configurations
+#------------------------------------------------------------------------------
+set(CMAKE_C_FLAGS ${CMAKE_C_FLAGS} -std=c99)
```
Comment (JaeseungYeom) [inline CMakeLists.txt:75]: We should just use the consistent standard with c++

Comment (hariharan-devarajan) [reply CMakeLists.txt:75]: when i switch the code to c++ we can move that right?

File: CMakeLists.txt (hunk 12 above)
Comment (hariharan-devarajan) [reply CMakeLists.txt]: Fixed

File: CMakeLists.txt (hunk 13 above)
Comment (hariharan-devarajan) [reply CMakeLists.txt]: FIXED

**Pull request merged**